app/src/main/java/com/github/orkest/Model/AudioRecognition.kt

```diff
@@ -0,0 +1,174 @@
+
+    companion object {
+
+         val catalog =
+                ShazamKit.createCustomCatalog().apply {
+
+                }
+
+        private fun Int.toByteAllocation(): Int {
+            return when (this) {
+                AudioFormat.ENCODING_PCM_16BIT -> 2
+                else -> throw IllegalArgumentException("Unsupported encoding")
+            }
+        }
+
+        private fun ByteBuffer.putTrimming(byteArray: ByteArray) {
+            if (byteArray.size <= this.capacity() - this.position()) {
+                this.put(byteArray)
+            } else {
+                this.put(byteArray, 0, this.capacity() - this.position())
+            }
+        }
+
+        /**
+         * Records audio for the maximum query duration and returns an [AudioChunk] with the recorded audio
+         * @return [AudioChunk] with the recorded audio
+         * @throws [SecurityException] if the app does not have the RECORD_AUDIO permission
+         */
+        @RequiresPermission(Manifest.permission.RECORD_AUDIO)
+        private fun audioChunkRecord(): AudioChunk {
+            val audioSource = MediaRecorder.AudioSource.UNPROCESSED
+
+            val audioFormat = AudioFormat.Builder()
+                .setChannelMask(AudioFormat.CHANNEL_IN_MONO)
+                .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
+                .setSampleRate(48_000)
+                .build()
+
+            val audioRecord = AudioRecord.Builder()
+                .setAudioSource(audioSource)
+                .setAudioFormat(audioFormat)
+                .build()
+
+            val seconds = catalog.maximumQuerySignatureDurationInMs
+
+            // Final desired buffer size to allocate necessary seconds of audio
+            val size = audioFormat.sampleRate * audioFormat.encoding.toByteAllocation() * seconds
+            val destination = ByteBuffer.allocate(size.toInt())
+
+            // Small buffer to retrieve chunks of audio
+            val bufferSize = AudioRecord.getMinBufferSize(
+                48_000,
+                AudioFormat.CHANNEL_IN_MONO,
+                AudioFormat.ENCODING_PCM_16BIT
+            )
+
+            audioRecord.startRecording()
+            val readBuffer = ByteArray(bufferSize)
+            while (destination.remaining()>0) {
+                val actualRead = audioRecord.read(readBuffer, 0, bufferSize)
+                val byteArray = readBuffer.sliceArray(0 until actualRead)
+                destination.putTrimming(byteArray)
+            }
+            audioRecord.release()
+
+            return AudioChunk(destination.array(), destination.position(), System.currentTimeMillis())
+        }
+
+        //TODO: Check if this is a correct way to do dependency injection
+        /**
+         * Records AudioChunks continuously and returns a [Flow] of [AudioChunk]s
+         * Must be called in the coroutineScope as the streaming session
+         */
+        @RequiresPermission(Manifest.permission.RECORD_AUDIO)
```

Comment:
This asks the user to allow the app to use the microphone ?

Reply:
No the permission still has to be requested beforehand, the annotation serves as a reminder that the permission must be set before calling this method to prevent runtime errors. The code actually doesn't compile without it.